teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +85,116 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+### Introduce “new-stable” value for enable-api-fields
+
+- A new option, `enable-api-fields` = `new-stable`, will be added to the API. This option will use the preferred validation where `stable` and `beta` features are validated the same across apiVersions. In 9 months, `new-stable` will be renamed as `stable`.
+- The current behavior for “stable” `enable-api-fields` remains the same for now where it allows users to use `beta` features in `v1beta1` that have coupled feature and api versioning in `v1beta1` i.e. remote resolution. In 9 months “stable” will be renamed as “legacy-stable” and be marked as deprecated.
+- The existing `beta` option will remain the same for now and in 9 months.
+
+- Taking the remote resolution feature, which currently couples feature and api versioning, as an example, after the change:
+  - With `new-stable` in `v1beta1`, we disable resolver.
+  - With `legacy-stable` in `v1beta1`, we do not validate resolvers as a field needs `enable-api-fields` as beta, so they are still turned on by default.
+  - When `enable-api-fields` is set to `beta` in both `v1` and `v1beta1`, we are turning on resolver.
+
+### Provide an immediate solution to decouple api and feature versioning now
+This solution proposes that we provide an immediate solution to turn on stable features only with `stable` `enable-api-fields` and require `enable-api-fields` to be set to `beta` when using beta features, regardless of CRD API version. This will solve the unintended behavior that taskruns cannot be created for version coupled features with “enable-api-fields” set to “stable” during the [v1 storage swap](https://github.com/tektoncd/pipeline/pull/6444#issuecomment-1580926707) right away.
+
+### Migrate `enable-api-fields` to `stable` in 9 months
+The default value for `enable-api-fields` will be switched back to `stable` in the long run, as discussed in [issue #6948](https://github.com/tektoncd/pipeline/issues/6948). The `new-stable` flag introduced will be renamed to `stable` in 9 months, at which point it will become the default value. And the current `stable` behavior will be renamed to `legacy-stable` for backwards compatibility. The `legacy-stable` flag will be deprecated and then removed when `v1beta1` is removed or once all current `beta` features that are coupled(object params and results, array results and indexing and remote resolution) become `stable`, whichever happens first.
+
+## Design Evalutaion
+### Pros
+- The proposed solution provides immediate options for users who want to use only stable features in `v1beta1`. This helps to prevent adding more complexity to the existing validation if more `beta` features are being added to `v1beta1`.
+- Workarounds are provided instead of having backwards incompatible changes during migration of the defaulting `enable-api-fields` back to `stable`. For example, users that accidentally have `beta` features opt-in with `v1beta1` CRDs could migrate to the usage with `legacy-stable` validation.
```

Comment:
Can you reword this? This change is backwards incompatible, we're just providing more warning and the option to retain legacy behavior.

Reply:
Updated! Thanks again for the clarifications on the backwards compatibility.